### ISSUE TITLE
Invalid json handling of empty objects {} converted to empty arrays []

### ISSUE THREAD

**Issue by joshribakoff** [FriendsOfSymfony/FOSRestBundle#980]
If you POST some json like this:

```
{foo:{}}
```

In my controller `$request->get('foo')` returns `[]` instead of `stdClass {}`. We're using Angular to POST to a controller that uses FOSRest.

This completely crashes our app because Angular cannot $watch an empty array.

I tried to debug & I found this:
```php
// FOS\RestBundle\Decoder\JsonDecoder
public function decode($data)
{
        return @json_decode($data, true);
}
```

From the PHP docs:
> mixed json_decode ( string $json [, bool $assoc = false [, int $depth = 512 [, int $options = 0 ]]] )
> assoc:
> When TRUE, returned objects will be converted into associative arrays.

This was kind of insidious for us to debug. I came up with this override which preserves my objects:

```
public function decode($data)
    {
        $obj = @json_decode($data);

        $newArr = [];
        foreach($obj as $key=>$value) {
            $newArr[$key] = $value;
        }
        return $newArr;
    }
```

However, certain symfony components  & commonly used 3rd party bundles still crash after this modification.

Currently, the only acceptable workaround we've found is to overwrite the corrupt json with valid json ourselves:
```
$request->get('foo); // "corrupted" by FOSRest
json_decode($request->getContent())['foo']; // this is the good stuff
```

**Comment by lsmith77** [comment]
I somehow seem to remember that we had this topic before .. at least 1-2 years past. note you can override the service used for decoding of json with a custom service.

**Comment by florianv** [comment]
We could maybe provide an option for the `JsonDecoder` to preserve objects ? 

The outermost object would still need to be encoded as associative array to be accessed using `$request->get('foo')`.

**Comment by lsmith77** [comment]
or we move to https://github.com/webmozart/json in case it already provides the options we need

/cc @webmozart

**Comment by webmozart** [comment]
If you set `$assoc` to `true`, that's what `json_decode()` does. You need to either set `$assoc` to `false` or convert empty arrays to objects manually. 

The problem is that `json_decode()` is not schema-aware, but neither is webmozart/json. You could maybe create a feature request on the PHP bug tracker?

**Comment by joshribakoff** [comment]
To be clear, this is not a PHP bug. PHP can preserve my data just fine.
```php
$str = "{foo:[], bar:{}}";
var_dump(json_encode(json_decode($str))) === $str; // true
```
json_decode() works correctly by default, `FOSRestBundle` is explicitly telling `json_decode()` to 'destroy' the data's schema. `json_decode` does not "destroy" data by default

Unfortunately I think the problem is more complex than simply modifying the JsonDecoder. `JMSSerializerBundle` also casts objects into arrays when serializing the response. Doctrine's "json array" property type also casts objects to arrays. When using the  `SyliusResourceBundle` (which depends on FOSRest) it passes the request to the Symfony form which also crashes if it gets an array not an object.

Hopefully you can see what I'm saying. Fixing FOSRest to not ask json_decode() to mangle the data isn't the problem. The problem is all the dependencies upon FOSRest that exist in the Symfony ecosystem that would break.

Casting objects to arrays it would seem is a PHP "standard" or convention. What I ended up doing was  casting the property back to an object in Angular side. It seems like a losing battle to try to convince all the PHP Bundles to change their ways. For our project, we'd need to override about 5 or more bundles, which seems pretty brittle & not future proof.

You can close this issue as expected behavior if you want.... I guess I'd just like to see it better documented. If I had known about "the PHP way" I'd have tried to work within that convention instead of spending time trying to override 5+ bundles (which wasn't working so well)

**Comment by joshribakoff** [comment]
I'm thinking out loud about a solution --

Authors of PHP libraries want to write `$foo['bar']` and not write `$foo->bar`, so they just cast all incoming json to an associative array. JS has no associative arrays, so PHP just casts non numerically indexed arrays to objects on the response.

This works -- JS sends an object, PHP converts to an array, processes it, converts it back to an object, JS gets an object back... Except it does not work, in the case of empty objects.... 

As an Angular developer, I want my backend to properly persist my JSON verbatim. Even for edge cases like empty objects. So what if my backend substituted empty objects with an "empty object" value object that serializes to an associative array? Confused?

Basically you need a request listener that:
- leaves the array [] as []
- converts the object {} to ['_blank_object'=>true]

Then you'd have a response listener that:
- leaves the array [] as []
- converts the array ['_blank_object'=>true] to a blank object {}

In theory, all the naughty PHP libraries can cast it to an array all they want without an issue now. They can remain ignorant as to the array's secret meaning.

**Comment by lsmith77** [comment]
maybe it would be safer to just add a flag on the request, like $request->attributes->set('_blank_object', true); which the view handler can then be made aware of

**Comment by mr-moon** [comment]
> json_decode() works correctly by default

```
$str = "{foo:[], bar:{}}";
var_dump(json_encode(json_decode($str)) === $str); // true
```

Actually, i'm getting false both with `PHP 5.6.30 (cli) (built: Aug  8 2017 12:20:45) ` and `PHP 7.0.22-0ubuntu0.16.04.1 (cli) ( NTS )` and `json version => 1.2.1` with `json version => 1.4.0` respectively.

Just FYI.

**Comment by cia05rf** [comment]
OK, i've just found this thread and i think the response is a simple one.
The documentation outlines how json_encode cna be used to force empty arrays to be converted to objects ([docs here](https://www.php.net/manual/en/function.json-encode.php#example-4696)).

It's as simple as:
`json_encode($b, JSON_FORCE_OBJECT)`

**Comment by dipankarjain** [comment]
@cia05rf  That would typecast even the arrays sent in the json payload to object. We would ideally want to preserve all types sent in the payload.
